fe/fe-core/src/main/java/org/apache/doris/planner/DecodeContext.java

```diff
@@ -0,0 +1,125 @@
+
+    private Map<Integer, Integer> slotIdToDictSlotId = new HashMap<>();
+
+    private Map<Integer, Integer> tupleIdToNewTupleWithDictSlot = new HashMap<>();
+
+    private boolean couldEncoded;
+
+    private Set<Integer> dictOptimizationDisabledSlot = new HashSet<>();
+
+    private Set<Integer> dictCodableSlotSet = new HashSet<>();
+
+    private DescriptorTable tableDescriptor;
+
+    private PlannerContext ctx_;
+
+    private Map<PlanNode, DecodeNode> childToDecodeNode = new HashMap<>();
+
+    public DecodeContext(PlannerContext ctx_, DescriptorTable tableDescriptor) {
+        this.ctx_ = ctx_;
+        this.tableDescriptor = tableDescriptor;
+    }
+
+    public void addDictCodableSlot(long tableId, int slotId) {
+        Set<Integer> slotSet = tableIdToDictCodableSlotSet.getOrDefault(tableId, new HashSet<>());
+        slotSet.add(slotId);
+        dictCodableSlotSet.add(slotId);
+    }
+
+    public void addAvailableDict(int slotId, ColumnDict dict) {
+        slotIdToColumnDict.put(slotId, dict);
+    }
+
+    public Set<Integer> getOriginSlotSet() {
+        return slotIdToColumnDict.keySet();
+    }
+
+    public Set<Integer> getAllDictCodableSlot() {
+        return dictCodableSlotSet;
+    }
+
+    public Set<Integer> getDictOptimizationDisabledSlot() {
+        return dictOptimizationDisabledSlot;
+    }
+
+    public boolean isCouldEncoded() {
+        return couldEncoded;
+    }
+
+    public void setCouldEncoded(boolean couldEncoded) {
+        this.couldEncoded = couldEncoded;
+    }
+
+    public TupleDescriptor generateTupleDesc(TupleId src) {
+        TupleDescriptor originTupleDesc = tableDescriptor.getTupleDesc(src);
+        TupleDescriptor tupleDesc = tableDescriptor.copyTupleDescriptor(src, "tuple-with-dict-slots");
+        tupleDesc.setTable(originTupleDesc.getTable());
+        tupleDesc.setRef(originTupleDesc.getRef());
+        tupleDesc.setAliases(originTupleDesc.getAliases_(), originTupleDesc.hasExplicitAlias());
+        tupleDesc.setCardinality(originTupleDesc.getCardinality());
+        tupleDesc.setIsMaterialized(originTupleDesc.getIsMaterialized());
+        tupleIdToNewTupleWithDictSlot.put(src.asInt(), tupleDesc.getId().asInt());
+        return tupleDesc;
+    }
+
+    public void addSlotToDictSlot(int slotId, int dictSlotId) {
+        slotIdToDictSlotId.put(slotId, dictSlotId);
+    }
+
+    public SlotDescriptor getDictSlotDesc(int slotId) {
+        int dictSlotId = slotIdToDictSlotId.get(slotId);
+        return tableDescriptor.getSlotDesc(new SlotId(dictSlotId));
+    }
+
+    public void updateSlotRefType(SlotRef slotRef) {
+        int slotId = slotRef.getSlotId().asInt();
+        SlotDescriptor dictSlotDesc = this.getDictSlotDesc(slotId);
+        slotRef.setDesc(dictSlotDesc);
+        slotRef.setType(Type.INT);
+    }
+
+    public DecodeNode newDecodeNode(PlanNode child, List<Integer> originSlotIdSet, ArrayList<TupleId> output) {
```

Comment:
Decode should come as late as possible. i think a better way do decide insert decode node is find the node should do function that need origin slot, and insert decode node before it.
So, we should find decode node's parent instead of child.

Reply:
You're right, but without project planner, it would be hard to decide the correct  required slot of parent. In the `AggNode::generateDecodeNode`, we leverage this feature by generating the DecodeNode according to `PlanNode::outputSlotIds` field first which is calculated by the required slot of its parent node.